src/perf_ringbuffer.cc

```diff
@@ -17,12 +18,23 @@ bool rb_init(RingBuffer *rb, void *base, size_t size,
   rb->data_size = size - rb->meta_size;
   rb->mask = get_mask_from_size(size);
   rb->type = ring_buffer_type;
-  if (ring_buffer_type == RingBufferType::kPerfRingBuffer) {
+  switch (ring_buffer_type) {
+  case RingBufferType::kPerfRingBuffer: {
     perf_event_mmap_page *meta =
         reinterpret_cast<perf_event_mmap_page *>(rb->base);
     rb->reader_pos = reinterpret_cast<uint64_t *>(&meta->data_tail);
     rb->writer_pos = reinterpret_cast<uint64_t *>(&meta->data_head);
-  } else {
+    break;
+  }
+  case RingBufferType::kMPSCRingBuffer: {
+    ddprof::MPSCRingBufferMetaDataPage *meta =
+        reinterpret_cast<ddprof::MPSCRingBufferMetaDataPage *>(rb->base);
+    rb->reader_pos = &meta->reader_pos;
+    rb->writer_pos = &meta->writer_pos;
+    rb->spinlock = &meta->spinlock;
```

Comment:
I'm kinda curious about this one -- why store pointers to the `MPSCRingBufferMetaDataPage` struct, rather than putting that struct directly inside `struct RingBuffer`? (potentially as a union, to support the multiple ring buffer types)

Reply:
I hesitated some time between both solutions. I think at the start I thought I could factorize code between ring buffer implementations by having a common pointer that expose the same semantic but points to different parts depending on the type.
But I agree that since I was not able to factorize code, clearly separating the data for different implementations might make the code mode readable.